APP CRASH (telock99-hostname.exe) single step execution

With version 6.2.0-2 of DynamoRio
The latest build does not solve the problem

On Windows 7, with a 32 bit application, telock99-hostname.exe the classic hostname software packed with telco

I run it with (no client)
C:\rio\bin32\drrun.exe -- antiRE.exe

The expected output is a line with the hostname.
Instead we get an application end.

The problem seems to come from a single step execution
```
TAG  0x01005d44
 +0    L3              9c                   pushf  %esp -> %esp 0xfffffffc(%esp)[4byte]
 +1    L3              81 0c 24 00 01 00 00 or     $0x00000100 (%esp)[4byte] -> (%esp)[4byte]
 +8    L3              75 02                jnz    $0x01005d50
END 0x01005d44
```
```
TAG  0x01005d50
 +0    L3              9d                   popf   %esp (%esp)[4byte] -> %esp
 +1    L3              eb 01                jmp    $0x01005d54
END 0x01005d50
```

The trap flag is set with pop instruction. 
We get an exception with ExceptionCode = 0x80000004 (EXCEPTION_SINGLE_STEP)
The problem is that ExceptionAddress is the EIP after the execution instruction , so it is the jump target
So the following condition is true in check_internal_exception in callback.c
```
            if (is_dynamo_address((app_pc)pExcptRec->ExceptionAddress) &&
                !in_fcache(pExcptRec->ExceptionAddress) ) {
```
because the jump target is not yet in fcache.

I guess that this condition should be modified, adding something like `&& pExcptRec->ExceptionCode != EXCEPTION_SINGLE_STEP`

Re-opening as 47b56c8 has two problems:

There are two problems with what was checked in:

1) It has several style violations caught by the not-yet-finished vera style checks (#2369, #2434):

```
/work/dr/git/src/core/win32/callback.c:5726: error line_length_content: Line too long: exceeds 90 chars.
/work/dr/git/src/core/dispatch.c:926: error line_length_content: Line too long: exceeds 90 chars.
/work/dr/git/src/core/arch/interp.c:823: error line_length_content: Line too long: exceeds 90 chars.
```

2) It breaks -tracedump_origins by causing a crash: 

http://dynamorio.org/CDash/testDetails.php?test=269690&build=23488

```
<Application /work/dr/nightly/run/build_debug-internal-32/suite/tests/bin/common.eflags (26115).  DynamoRIO internal crash at PC 0x567cb888.  Please report this at http://dynamorio.org/issues/.  Program aborted.
Received SIGSEGV at pc 0x567cb888 in thread 26115
Base: 0x565b1000
Registers:eax=0xffffffff ebx=0x568fb388 ecx=0x00000000 edx=0xef7c7658
	esi=0x00006603 edi=0x568fb388 esp=0xef7928c0 ebp=0xef792a14
	eflags=0x00010246
```

```
Program received signal SIGSEGV, Segmentation fault.
0x000055b71361a798 in build_bb_ilist (dcontext=0xffffffffffffffff, bb=0x4b21e500) at /work/dr/git/src/core/arch/interp.c:3381
3381        if (dcontext->single_step_addr == bb->start_pc) {
(gdb) bt
#0  0x000055b71361a798 in build_bb_ilist (dcontext=0xffffffffffffffff, bb=0x4b21e500) at /work/dr/git/src/core/arch/interp.c:3381
#1  0x000055b713620b34 in build_app_bb_ilist (dcontext=0xffffffffffffffff, 
    start_pc=0x7f82729b79a0 "L\213{\bE\211\375I\203\375%\017\204\357\003", outf=4000) at /work/dr/git/src/core/arch/interp.c:4555
#2  0x000055b71360d57d in disassemble_app_bb (dcontext=0xffffffffffffffff, 
    tag=0x7f82729b79a0 "L\213{\bE\211\375I\203\375%\017\204\357\003", outfile=4000)
    at /work/dr/git/src/core/arch/disassemble_shared.c:1417
#3  0x000055b71344ba5c in output_trace (dcontext=0xffffffffffffffff, pt=0x4b18b4f8, f=0x4b2969e0, deleted_at=-1)
    at /work/dr/git/src/core/fragment.c:7281
#4  0x000055b713435f37 in fragment_exit () at /work/dr/git/src/core/fragment.c:1604
```

--- a/core/arch/interp.c
+++ b/core/arch/interp.c
@@ -820,7 +820,8 @@ check_new_page_jmp(dcontext_t *dcontext, build_bb_t *bb, app_pc new_pc)
 static inline void
 bb_process_single_step(dcontext_t *dcontext, build_bb_t *bb)
 {
-    LOG(THREAD, LOG_INTERP, 2, "interp: single step exception bb at "PFX"\n", bb->instr_start);
+    LOG(THREAD, LOG_INTERP, 2, "interp: single step exception bb at "PFX"\n",
+        bb->instr_start);
     /* FIXME i#2144 : handling a rep string operation.
      * In this case, we should test if only one iteration is done
      * before the single step exception.
@@ -3377,7 +3378,7 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
         bb->full_decode = true;
         bb->record_translation = true;
     }
-    if (dcontext->single_step_addr == bb->start_pc) {
+    if (my_dcontext != NULL && my_dcontext->single_step_addr == bb->start_pc) {
         /* Decodes only one instruction because of single step exception. */
         cur_max_bb_instrs = 1;
     }
@@ -3712,7 +3713,7 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
 # endif /* X86 */
 #endif /* WINDOWS */
 
-        if (dcontext->single_step_addr == bb->start_pc) {
+        if (my_dcontext != NULL && my_dcontext->single_step_addr == bb->start_pc) {
             bb_process_single_step(dcontext, bb);
             /* Stops basic block right now. */
             break;
@@ -4203,7 +4204,7 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
 #ifdef HOT_PATCHING_INTERFACE
         && !hotp_injected
 #endif
-        && dcontext->single_step_addr != bb->start_pc
+        && (my_dcontext == NULL || my_dcontext->single_step_addr != bb->start_pc)
        ) {
         /* If the fragment doesn't have a syscall or contains a
          * non-ignorable one -- meaning that the frag will exit the cache
@@ -4233,7 +4234,7 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
         }
 #endif
     }
-    else if (dcontext->single_step_addr == bb->start_pc) {
+    else if (my_dcontext != NULL && my_dcontext->single_step_addr == bb->start_pc) {
         /* Field exit_type might have been cleared by client_process_bb. */
         bb->exit_type |= LINK_SPECIAL_EXIT;
     }

--- a/core/dispatch.c
+++ b/core/dispatch.c
@@ -923,7 +923,8 @@ dispatch_enter_dynamorio(dcontext_t *dcontext)
                 STATS_INC(float_pc_from_dispatch);
                 /* Restore */
                 dcontext->upcontext.upcontext.exit_reason = EXIT_REASON_SELFMOD;
-            } else if (dcontext->upcontext.upcontext.exit_reason == EXIT_REASON_SINGLE_STEP) {
+            } else if (dcontext->upcontext.upcontext.exit_reason ==
+                       EXIT_REASON_SINGLE_STEP) {
                 /* Delete basic block to generate only one single step exception. */
                 ASSERT(!TEST(FRAG_SHARED, dcontext->last_fragment->flags));
                 fragment_delete(dcontext, dcontext->last_fragment, FRAGDEL_ALL);

--- a/core/win32/callback.c
+++ b/core/win32/callback.c
@@ -5723,7 +5723,8 @@ intercept_exception(app_state_at_intercept_t *state)
 #endif
                         }
                         instr_free(dcontext, &instr);
-                        transfer_to_dispatch(dcontext, &mcontext, false/*!full_DR_state*/);
+                        transfer_to_dispatch(dcontext, &mcontext,
+                                             false/*!full_DR_state*/);
                         ASSERT_NOT_REACHED();
                     }
                 }

